Script to create a stable build for ICDS

emord opened the pull request:
Saw that ICDS will eventually get added to the normal deploy rotation on Mondays. This week seems like it has been particularly bad with many reverts/hotfixes happening after deploy, which makes this feel even more important. This would hopefully cut down on high visibility bugs hitting ICDS and stressful IST firefighting.

My proposal for this is to base the branch on the tag from the Friday deploy and merge specific PRs into that branch. (We can't use branch names as they are often deleted after merging). I'm assuming that between Friday and Monday any issues will have been identified and we can get specific PRs in before then, but could also change to Thursday if we want a full working day.

Followup if going this route:
- [ ] change ICDS to always deploy enterprise_stable branch
- [ ] change other enterprise deployments (swiss/PNA) to always deploy this branch
- [ ] automatically get the last tag that was deployed to production (should probably always be last friday though)
- [ ] is there a way to warn if one of the branches contains a migration?

This is a super hacky 30 min MVP but when I was glancing through rebuildstaging it looked like random hacks were par for the course.  this script could probably use some love in general.

@dimagi/scale-team  Thoughts on this?
@dannyroberts you'll probably have good thoughts on this as well

emord committed:
Script to create a stable build

czue commented:
> Thoughts on this?

The process sounds good to me and I think is a good mix of not-too-delayed, but not-too-volatile. 

>  automatically get the last tag that was deployed to production (should probably always be last friday though)

Is the idea that we're relying on a manual process that builds `enterprise_stable` from the last production-deployed branch sometime between the friday and monday deploys? Sounds reasonable, though also possibly a bit brittle if there were weekend/Monday hotfixes. 

>  change ICDS to always deploy enterprise_stable branch

I think we probably want to also add some checks baked into the deploy process that monitor the last time it was built. Like if I run `fab icds deploy` and didn't rebuild `enterprise_stable` it should somehow trip that I am deploying week-old code. Likewise, it might be worth flagging if the prod base commit is too new?

snopoke commented:
We've discussed the option of having a [stable branch](https://docs.google.com/document/d/1omagW_sGa0cRBziP3T9gY3HTZfb7R9VHzeycNcv-q24/edit#) before but got some push back (see doc for details). I do like that this follows a similar workflow to staging.

Maybe rather than taking the Friday tag we should take a tag from mid week that's had more time to 'become stable'.

* What happens when a PR gets merged? Does the git reference still exist?
* For any bugs that happen after the 'prod' deploy tag, who would be responsible for making sure they go into the stable 'spec'?

emord commented:
> Sounds reasonable, though also possibly a bit brittle if there were weekend/Monday hotfixes.

I think that the idea would be if anything needs to be hotfixed, it needs to be put into this file under `pull_requests` as it would be very out of the ordinary

> Like if I run fab icds deploy and didn't rebuild enterprise_stable it should somehow trip that I am deploying week-old code. Likewise, it might be worth flagging if the prod base commit is too new?

Agreed on both these checks

> What happens when a PR gets merged? Does the git reference still exist?

The ref to the PR still exists. I checked with a couple of PRs with delete branches

> For any bugs that happen after the 'prod' deploy tag, who would be responsible for making sure they go into the stable 'spec'?

I'm not sure. We could add to our "whoa there bud" checks that says "Have you rebuilt enterprise_stable and gone through (list of PRs merged/created since last tag) to see what should be included?" It will likely be forgotten to be added when the hotfix happens, so forcing a check on deploy would likely be the best way. Any deploys that happen after Monday, would be for specific bugs so it seems less likely that we would forget the hotfix.

> We've discussed the option of having a stable branch before but got some push back (see doc for details). I do like that this follows a similar workflow to staging.

Yeah I recall us discussing it before, but think we should re-open the discussion. From the feedback at the bottom of the [doc](https://docs.google.com/document/d/1omagW_sGa0cRBziP3T9gY3HTZfb7R9VHzeycNcv-q24/edit#), it seems like the hesitation was on calling/referring to this as a  "stable" release. I agree with @orangejenny @millerdev that this is not really "stable" and happy to call it something else (maybe just drop stable and call it the "enterprise" branch. 

> Maybe rather than taking the Friday tag we should take a tag from mid week that's had more time to 'become stable'.

That seems fine to me, but I think it's more likely to have an issue as mentioned in the doc by @orangejenny :

> There's also potentially some weirdness with someone releasing feature A, and then someone else does bugfixes in the first week that touch that feature

emord commented:
To be totally clear with my thought on this process it would be:

1. Deploy production from latest master on Friday
2. Some subset of users will be using it over the weekend and throughout Monday
3. If nothing major has been reported by Monday ~3 PM EST, deploy Friday's deploy to ICDS
4. If something was reported and has been fixed/hotfixed to prod, deploy Friday's deploy + that branch to ICDS, preferably also running the full test suite on the "enterprise" branch before deploying
5. During the week, if an issue has been found/reported on ICDS, PR a fix as normal, add it to the enterprise branch and deploy to ICDS. If the issue has not been reported on ICDS or isn't a high priority bug fix, do nothing.

From my perspective this should eliminate any P1/P2 bugs that aren't custom ICDS code and we should have slightly higher confidence in the code than just deploying directly from master.

I think that this provides the best compromise that I see as feasible in the short term, but would like to here @orangejenny and @millerdev thoughts on that to see if their original concerns are sufficiently addressed

dannyroberts commented:
I'm a little late to the party, but I just wanted to say that if we did decide to go with a process like this, I like the approach. It also occurs to me that this rebuildstaging workflow is a pretty general thing that others could use. Could be worth pulling out and adding to our little pile of open source libraries.

This also reminds me that when others are managing their own commcare clusters, we'll probably need to have some kind of versioning/stable release system.

orangejenny commented:
Offlined with Emord, seems fine to me.

millerdev commented:
> automatically get the last tag that was deployed to production (should probably always be last friday though)

This sounds good to me. If there was a hotfix deploy since Friday then we would use that instead, correct?

In other words, on Monday we deploy to ICDS whatever has been running on production since Friday (or a subsequent hotfix if applicable).

millerdev commented:
+1 to calling it "enterprise" rather than "stable"

emord committed:
Remove references to stable

emord committed:
Reorder imports

emord committed:
Fix BranchConfig definition

emord committed:
Add util to get the most recent tags

emord committed:
Check that the trunk is recent if it's a tag

emord committed:
Update enterprise yaml

emord commented:
I removed all references to "stable" here so that it's clear it's just enterprise. This hits at the awkward spot between commcare-hq and commcare-cloud, so I split the responsibilites like this:

commcare-cloud [defaults to deploying the "enterprise" branch](https://github.com/dimagi/commcare-cloud/pull/1740/commits/2286cf84f5ae3cdbbdbc79f6767fd4d095e21bd0) to our enterprise environments
commcare-cloud warns if we are deploying the [same code as the previous deploy](https://github.com/dimagi/commcare-cloud/pull/1740/commits/09b90e732cd5d3ffd1444302e92e46de0aa4d164) (which is what would happen if you forgot to rebuild the branch)

`rebuildstaging` has a new option `--enterprise` (bad naming but wanted to keep the same workflow)
if `trunk` is a tag, it will warn you if it's not from the [latest 10 deploys and then suggests some tags to use](https://github.com/dimagi/commcare-hq/pull/20354/commits/dd9c77ae614760be7979defa341b2bf2cd2bf72e)

Not quite automatic, but hopefully a reasonable compromise on stability/flexibility
